docs/assignment3.md

```diff
@@ -0,0 +1,503 @@
+        when there were no different environments, the <b>Scene</b> class implemented the <b>Interactable</b> interface directly, but with more environments added, there
+        needed to be a "middle man", not only to have these multiple classes call similar methods but also to keep track of which environment the game is in, and how to
+        get back to the environment it was previously in.</td>
+        <td>The solution to this problem was the Decorator design pattern. This design pattern allows for objects to be modified at run time, a.k.a. having the
+        capabilities of inheritance at run-time. This was done in the implementation using an enumerator for the types of effects that can occur: a
+        navigation effect, a stats effect, and an inventory effect.</td>
+    </tr>
+    <tr>
+        <td><b>Intended use</b></td>
+        <td>When a command is entered by the player, the *handleCommand(Command, Callback)* method is called on the current <b>View</b>. The current View then either handles this command, or finds the next <b>Interactable</b> that knows how to handle this command. (Again by calling *handleCommand()*). This Interactable can again pass the command over to the next one in the chain.</td>
+        <td>The <b>Game</b> object keeps track of attached <b>Client</b> instances, and sends them information about what is happening in the game. This method is applied in other areas as well.</td>
+        <td>At run-time, the player would, for example, engage in combat with another character in the game. Once this happens, the method that controls entering the new "view" or environment is called and the combat scene begins, with the previous,
+        non-combat scene being stored for later. Once the player exits combat, the method to exit the environment is called from the abstract class, and then the previous environment is replaced, so the player
+        is back to where they were in the game before engaging in combat. </td>
+        <td>At run-time, when a player enters an action command, a method is called that returns a list of objects of class <b>Effect</b>. These objects have
+        their own descriptions and a type. Then, the type is the only thing that is important, and the appropriate changes are made depending on the
+        type of effect from the enum variables, and the appropriate output is made to the player on the CLI if it is appropriate.</td>
+    </tr>
+    <tr>
+        <td><b>Constraints</b></td>
+        <td>N/A</td>
+        <td>N/A</td>
+        <td>N/A</td>
+        <td>N/A</td>
+    </tr>
+    <tr>
+        <td><b>Additional remarks</b></td>
+        <td>N/A</td>
+        <td>The observable pattern is not fully implemented throughout our system.</td>
+        <td>N/A</td>
+        <td>N/A</td>
+    </tr>
+</table>
+
+Word Count: 840
+
```

Comment:
Well thought application of design patterns, Great description of the rationale for applying them.
Note: Template is implemented to solve a rather niche problem, could have been used for character-enemy-npc instead.